src/main/java/edu/ucla/library/libcal/services/OAuthTokenServiceImpl.java

```diff
@@ -109,10 +101,59 @@ protected Future<Void> shareAccessToken(final User aToken) {
         });
     }
 
-    @Override
-    public Future<Void> close() {
-        myVertx.cancelTimer(myTimerId);
+    /**
+     * Refresh the OAuth token before it has a chance to expire. This method is recursive.
+     *
+     * @param aToken The OAuth token
+     * @param aSecondsBeforeExpiration The number of seconds before expiration to attempt a refresh
+     * @return The ID of a Vert.x timer
+     */
+    private long keepTokenFresh(final User aToken, final int aSecondsBeforeExpiration) {
+        final int delay = aToken.principal().getInteger(JsonKeys.EXPIRES_IN) - aSecondsBeforeExpiration;
 
-        return Future.succeededFuture();
+        return myVertx.setTimer(delay, timerID -> {
+            authenticateWithRetry(Optional.of(3), 5).compose(newToken -> {
```

Comment:
Here, I have the number of refresh attempts hard-coded to 3, but we could make this configurable (and could instead pass `Optional.empty()` for unlimited attempts).